### PR TITLE
Fixes #2658

### LINKED ISSUE #2658: Product flat tables broken, ref #2603
### Preconditions (*)

1. 19.4.x-latest
2. Sample Data installed
3. Flat product tables ON

### Steps to reproduce (*)

1. ge to https://openmage.ddev.site/men/new-arrivals.htm

### Actual result (*)

```
SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '1=1 AND ( e.website_id = '1' ) AND ( e.customer_group_id = 0) AND (e.min_pric...' at line 2, query was: SELECT FLOOR((ROUND((e.min_price) * 1, 2)) / 100) + 1 AS `range`, COUNT(*) AS `count` FROM `catalog_product_index_price` AS `e`
 INNER JOIN `catalog_category_product_index` AS `cat_index` ON cat_index.product_id=e.entity_id AND cat_index.store_id=3 AND cat_index.visibility IN(2, 4) AND cat_index.category_id = '14' WHERE 1=1 1=1 AND ( e.website_id = '1' ) AND ( e.customer_group_id = 0) AND (e.min_price IS NOT NULL) GROUP BY FLOOR((ROUND((e.min_price) * 1, 2)) / 100) + 1 ORDER BY FLOOR((ROUND((e.min_price) * 1, 2)) / 100) + 1 ASC

#0 /var/www/html/lib/Varien/Db/Statement/Pdo/Mysql.php(104): Zend_Db_Statement_Pdo->_execute(Array)
#1 /var/www/html/app/code/core/Zend/Db/Statement.php(289): Varien_Db_Statement_Pdo_Mysql->_execute(Array)
#2 /var/www/html/lib/Zend/Db/Adapter/Abstract.php(480): Zend_Db_Statement->execute(Array)
#3 /var/www/html/lib/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract->query('SELECT FLOOR((R...', Array)
#4 /var/www/html/lib/Varien/Db/Adapter/Pdo/Mysql.php(502): Zend_Db_Adapter_Pdo_Abstract->query('SELECT FLOOR((R...', Array)
#5 /var/www/html/lib/Zend/Db/Adapter/Abstract.php(811): Varien_Db_Adapter_Pdo_Mysql->query('SELECT FLOOR((R...', Array)
#6 /var/www/html/app/code/core/Mage/Catalog/Model/Resource/Layer/Filter/Price.php(272): Zend_Db_Adapter_Abstract->fetchPairs(Object(Varien_Db_Select))
#7 /var/www/html/app/code/core/Mage/Catalog/Model/Layer/Filter/Price.php(152): Mage_Catalog_Model_Resource_Layer_Filter_Price->getCount(Object(Mage_Catalog_Model_Layer_Filter_Price), 100)
```

See ` WHERE 1=1 1=1` ... related code ... `Mage_Catalog_Model_Resource_Layer_Filter_Price::_getSelect()`

```
        // processing WHERE part
        $wherePart = $select->getPart(Zend_Db_Select::WHERE);
        $excludedWherePart = Mage_Catalog_Model_Resource_Product_Collection::MAIN_TABLE_ALIAS . '.status';
        foreach ($wherePart as $key => $wherePartItem) {
            if (strpos($wherePartItem, $excludedWherePart) !== false) {
                $wherePart[$key] = new Zend_Db_Expr('1=1');
                continue;
            }
            $wherePart[$key] = $this->_replaceTableAlias($wherePartItem);
        }
```

Before #2603 

```
" FROM `catalog_product_index_price` AS `e`
 INNER JOIN `catalog_category_product_index` AS `cat_index` ON cat_index.product_id=e.entity_id AND cat_index.store_id=3 AND cat_index.visibility IN(2, 4) AND cat_index.category_id = '14' WHERE 1=1 AND ( e.website_id = '1' ) AND ( e.customer_group_id = 0) AND (e.min_price IS NOT NULL)"
```

```
$wherePart = {object[1]} 
 0 = {Zend_Db_Expr} "1=1"
$wherePartItem = "(e.status = 1)"
```

After:

```
" FROM `catalog_product_index_price` AS `e`
 INNER JOIN `catalog_category_product_index` AS `cat_index` ON cat_index.product_id=e.entity_id AND cat_index.store_id=3 AND cat_index.visibility IN(2, 4) AND cat_index.category_id = '14' WHERE 1=1 1=1 AND ( e.website_id = '1' ) AND ( e.customer_group_id = 0) AND (e.min_price IS NOT NULL)"
```

```
$wherePart = {object[2]} 
 0 = {Zend_Db_Expr} "1=1"
 1 = {Zend_Db_Expr} "1=1"
$wherePart = {string[2]} ["(e.status = 1)", "AND (e.status I..."]
 0 = "(e.status = 1)"
 1 = "AND (e.status IN(1))"
```

### PR DESCRIPTION
**Pull request by sreichel**
### Description (*)

#2603 was for filter disabled products in admin backend / create orders. Some changes have inpact on frontend and should be reverted.

### Related Pull Requests

1. Fixes OpenMage/magento-lts#2603

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#2658
2. Closes OpenMage/magento-lts#2659
3. Closes OpenMage/magento-lts#2660

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list

### CODE REVIEW

File: app/code/core/Mage/Catalog/Model/Layer.php [hunk 1]
```diff
@@ -115,10 +115,7 @@ public function prepareProductCollection($collection)
             ->addAttributeToSelect(Mage::getSingleton('catalog/config')->getProductAttributes())
             ->addPriceData()
             ->addTaxPercents()
-            ->addUrlRewrite($this->getCurrentCategory()->getId())
-            ->addAttributeToFilter('status', [
-                'in' => Mage::getSingleton('catalog/product_status')->getVisibleStatusIds()
-            ]);
```
Comment (fballiano) [inline app/code/core/Mage/Catalog/Model/Layer.php:121]: I don't think this code should be removed, the previous version of this file had

`Mage::getSingleton('catalog/product_status')->addVisibleFilterToCollection($collection);`

which was filtering the status (although it wasn't working)

Comment (sreichel) [reply app/code/core/Mage/Catalog/Model/Layer.php:121]: What is the benefit of your change?

There was no status filter before. Why to add this filter, that is removed later?

Comment (fballiano) [reply app/code/core/Mage/Catalog/Model/Layer.php:121]: maybe I'm missing something

Comment (sreichel) [reply app/code/core/Mage/Catalog/Model/Layer.php:121]: `addVisibleFilterToCollection()` did nothing (it just returns `$this`) ... so there was no filter on `status`. It was never used, and marked as deprecated from beginning ... maybe for some reason? Maybe b/c it was implemented in another way?

Your change to `Catalog/Model/Layer.php` added a filter that gets removed here:

```
 $excludedWherePart = Mage_Catalog_Model_Resource_Product_Collection::MAIN_TABLE_ALIAS . '.status'; 
 foreach ($wherePart as $key => $wherePartItem) { 
     if (strpos($wherePartItem, $excludedWherePart) !== false) { 
         $wherePart[$key] = new Zend_Db_Expr('1=1'); 
         continue; 
     } 
     $wherePart[$key] = $this->_replaceTableAlias($wherePartItem); 
 } 
```

For enabled flat product tables, status filter is set in `Mage_Catalog_Model_Resource_Product_Collection::_initSelect()` and thats gets replaced by `1=1` as first part of the WHERE condition. Nothing wrong so far. `1=1` is no ugly hack, its a comon pattern for dynamically created queriers, b/c you can add `AND/OR [condition]` w/o to care if it is first part in WHERE.

Also `setPart()` works. There is nothing lost. Before #2603 we had these part `"(e.status = 1)"`. After your change we have 

```
 0 = "(e.status = 1)"
 1 = "AND (e.status IN(1))" # this comes from your change
```

The "where type" `AND` is there (nothing that can get lost), but both are replaced with `1=1`, leading to wrong query.

I agree ... this is wrong (or at least not planned and covered to have `status` multiple times).

Again ...every  `e.status` gets removed intentionally, so it makes no sense to (re-)add it with `addAttributeToFilter('status', [ ....`

**Review by fballiano — APPROVED** [review]

**Review by ADDISON74 — APPROVED** [review]

**Pull request merged**